File: static/typescript/binary_ref.ts
// PolyGen indexed binary document support for TypeScript.
// Mirrors the C# BinaryRef format: eager indexes, lazy row field access.

const MAGIC = new Uint8Array([0x50, 0x47, 0x42, 0x52, 0x45, 0x46, 0x31, 0x00]); // PGBREF1\0
const VERSION = 1;
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

export class BinaryDocumentOwner {
  readonly bytes: Uint8Array;
  readonly view: DataView;

  constructor(input: ArrayBuffer | Uint8Array) {
    this.bytes = input instanceof Uint8Array ? input : new Uint8Array(input);
    this.view = new DataView(this.bytes.buffer, this.bytes.byteOffset, this.bytes.byteLength);
  }

  get length(): number {
    return this.bytes.byteLength;
  }

  reader(offset = 0): BinaryRefReader {
    return new BinaryRefReader(this.bytes, offset);
  }
}

export class BinaryRefReader {
  readonly bytes: Uint8Array;
  readonly view: DataView;
  offset: number;

  constructor(input: ArrayBuffer | Uint8Array, offset = 0) {
    this.bytes = input instanceof Uint8Array ? input : new Uint8Array(input);
    this.view = new DataView(this.bytes.buffer, this.bytes.byteOffset, this.bytes.byteLength);
    this.offset = offset;
    BinaryRefFormat.checkRange(this.bytes, offset, 0);
  }

  position(): number {
    return this.offset;
  }

  seek(offset: number): void {
    BinaryRefFormat.checkRange(this.bytes, offset, 0);
    this.offset = offset;
  }

  skip(length: number): void {
    BinaryRefFormat.checkRange(this.bytes, this.offset, length);
    this.offset += length;
  }

  readBool(): boolean {
    return this.readU8() !== 0;
  }

  readU8(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 1);
    return this.bytes[this.offset++];
  }

  readI8(): number {
    const value = this.readU8();
    return value > 127 ? value - 256 : value;
  }

  readU16(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 2);
    const value = this.view.getUint16(this.offset, true);
    this.offset += 2;
    return value;
  }

  readI16(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 2);
    const value = this.view.getInt16(this.offset, true);
    this.offset += 2;
    return value;
  }

  readU32(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 4);
    const value = this.view.getUint32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readI32(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 4);
    const value = this.view.getInt32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readU64(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 8);
    const value = Number(this.view.getBigUint64(this.offset, true));
    this.offset += 8;
    return value;
  }

  readI64(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 8);
    const value = Number(this.view.getBigInt64(this.offset, true));
    this.offset += 8;
    return value;
  }

  readF32(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 4);
    const value = this.view.getFloat32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readF64(): number {
    BinaryRefFormat.checkRange(this.bytes, this.offset, 8);
    const value = this.view.getFloat64(this.offset, true);
    this.offset += 8;
    return value;
  }

  readBytes(length: number): Uint8Array {
    BinaryRefFormat.checkRange(this.bytes, this.offset, length);
    const value = this.bytes.subarray(this.offset, this.offset + length);
    this.offset += length;
    return value;
  }
}

export class BinaryRefWriter {
  private chunks: Uint8Array[] = [];
  private totalLength = 0;

  get length(): number {
    return this.totalLength;
  }

  writeBool(value: boolean): void {
    this.writeU8(value ? 1 : 0);
  }

  writeU8(value: number): void {
    this.push(new Uint8Array([value & 0xff]));
  }

  writeI8(value: number): void {
    this.writeU8(value);
  }

  writeU16(value: number): void {
    const b = new Uint8Array(2);
    new DataView(b.buffer).setUint16(0, value, true);
    this.push(b);
  }

  writeI16(value: number): void {
    const b = new Uint8Array(2);
    new DataView(b.buffer).setInt16(0, value, true);
    this.push(b);
  }

  writeU32(value: number): void {
    const b = new Uint8Array(4);
    new DataView(b.buffer).setUint32(0, value, true);
    this.push(b);
  }

  writeI32(value: number): void {
    const b = new Uint8Array(4);
    new DataView(b.buffer).setInt32(0, value, true);
    this.push(b);
  }

  writeU64(value: number): void {
    const b = new Uint8Array(8);
    new DataView(b.buffer).setBigUint64(0, BigInt(value), true);
    this.push(b);
  }

  writeI64(value: number): void {
    const b = new Uint8Array(8);
    new DataView(b.buffer).setBigInt64(0, BigInt(value), true);
    this.push(b);
  }

  writeF32(value: number): void {
    const b = new Uint8Array(4);
    new DataView(b.buffer).setFloat32(0, value, true);
    this.push(b);
  }

  writeF64(value: number): void {
    const b = new Uint8Array(8);
    new DataView(b.buffer).setFloat64(0, value, true);
    this.push(b);
  }

  writeRaw(bytes: Uint8Array): void {
    this.push(bytes);
  }

  toUint8Array(): Uint8Array {
    const out = new Uint8Array(this.totalLength);
    let cursor = 0;
    for (const chunk of this.chunks) {
      out.set(chunk, cursor);
      cursor += chunk.byteLength;
    }
    return out;
  }

  private push(bytes: Uint8Array): void {
    this.chunks.push(bytes);
    this.totalLength += bytes.byteLength;
  }
}

export class BinaryRefFormat {
  static writeHeader(writer: BinaryRefWriter): void {
    writer.writeRaw(MAGIC);
    writer.writeI32(VERSION);
  }

  static readHeader(reader: BinaryRefReader): void {
    const magic = reader.readBytes(MAGIC.byteLength);
    for (let i = 0; i < MAGIC.byteLength; i++) {
      if (magic[i] !== MAGIC[i]) {
        throw new Error("Invalid PolyGen binary ref header.");
      }
    }
    const version = reader.readI32();
    if (version !== VERSION) {
      throw new Error(`Unsupported PolyGen binary ref version: ${version}.`);
    }
  }

  static writeString(writer: BinaryRefWriter, value: string): void {
    const bytes = textEncoder.encode(value);
    writer.writeI32(bytes.byteLength);
    writer.writeRaw(bytes);
  }

  static readString(reader: BinaryRefReader): string {
    const length = reader.readI32();
    if (length < 0) throw new Error("Negative string length.");
    return textDecoder.decode(reader.readBytes(length));
  }

  static writeBytes(writer: BinaryRefWriter, value: Uint8Array): void {
    writer.writeI32(value.byteLength);
    writer.writeRaw(value);
  }

  static readLengthPrefixedBytes(buffer: Uint8Array, offset: number): Uint8Array {
    this.checkRange(buffer, offset, 4);
    const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
    const length = view.getInt32(offset, true);
    if (length < 0) throw new Error("Negative binary payload length.");
    const payloadOffset = offset + 4;
    this.checkRange(buffer, payloadOffset, length);
    return buffer.subarray(payloadOffset, payloadOffset + length);
  }

  static readUtf8String(buffer: Uint8Array, offset: number): string {
    return textDecoder.decode(this.readLengthPrefixedBytes(buffer, offset));
  }

  static requireFieldOffset(buffer: Uint8Array, rowOffset: number, fieldIndex: number): number {
    const offset = this.getFieldOffset(buffer, rowOffset, fieldIndex);
    if (offset < 0) throw new Error(`Missing required binary field at index ${fieldIndex}.`);
    return offset;
  }

  static getFieldOffset(buffer: Uint8Array, rowOffset: number, fieldIndex: number): number {
    this.checkRange(buffer, rowOffset, 4);
    const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
    const fieldCount = view.getInt32(rowOffset, true);
    if (fieldCount < 0) throw new Error("Negative binary field count.");
    if (fieldIndex < 0 || fieldIndex >= fieldCount) return -1;

    const tableOffset = rowOffset + 4;
    this.checkRange(buffer, tableOffset, fieldCount * 4);
    const relative = view.getInt32(tableOffset + fieldIndex * 4, true);
    if (relative < 0) return -1;
    const absolute = rowOffset + relative;
    this.checkRange(buffer, absolute, 0);
    return absolute;
  }

  static readBool(buffer: Uint8Array, offset: number): boolean {
    this.checkRange(buffer, offset, 1);
    return buffer[offset] !== 0;
  }

  static readU8(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 1);
    return buffer[offset];
  }

  static readI8(buffer: Uint8Array, offset: number): number {
    const value = this.readU8(buffer, offset);
    return value > 127 ? value - 256 : value;
  }

  static readU16(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 2);
    return new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getUint16(offset, true);
  }

  static readI16(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 2);
    return new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getInt16(offset, true);
  }

  static readU32(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 4);
    return new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getUint32(offset, true);
  }

  static readI32(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 4);
    return new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getInt32(offset, true);
  }

  static readU64(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 8);
    return Number(new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getBigUint64(offset, true));
  }

  static readI64(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 8);
    return Number(new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getBigInt64(offset, true));
  }

  static readF32(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 4);
    return new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getFloat32(offset, true);
  }

  static readF64(buffer: Uint8Array, offset: number): number {
    this.checkRange(buffer, offset, 8);
    return new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength).getFloat64(offset, true);
  }

  static readTimestamp(buffer: Uint8Array, offset: number): Date {
    const ticks = this.readI64(buffer, offset);
    return new Date(ticks / 10000 - 62135596800000);
  }

  static writeTimestamp(writer: BinaryRefWriter, value: Date): void {
    writer.writeI64((value.getTime() + 62135596800000) * 10000);
  }

  static checkRange(buffer: Uint8Array, offset: number, length: number): void {
    if (offset < 0 || length < 0 || offset > buffer.byteLength || length > buffer.byteLength - offset) {
      throw new Error("Binary ref offset is outside the document.");
    }
  }
}

export class BinaryRefRowBuilder {
  private readonly fields: Array<Uint8Array | null>;

  constructor(fieldCount: number) {
    if (fieldCount < 0) throw new Error("Negative binary field count.");
    this.fields = new Array(fieldCount).fill(null);
  }

  setField(index: number, write: (writer: BinaryRefWriter) => void): void {
    const writer = new BinaryRefWriter();
    write(writer);
    this.fields[index] = writer.toUint8Array();
  }

  toUint8Array(): Uint8Array {
    const writer = new BinaryRefWriter();
    writer.writeI32(this.fields.length);

    let cursor = 4 + this.fields.length * 4;
    for (const field of this.fields) {
      if (field == null) {
        writer.writeI32(-1);
      } else {
        writer.writeI32(cursor);
        cursor += field.byteLength;
      }
    }

    for (const field of this.fields) {
      if (field != null) writer.writeRaw(field);
    }

    return writer.toUint8Array();
  }
}

File: static/csharp/BinaryRef.cs
// This file is a part of the Polygen common utility library.
// It provides indexed binary document support for lazy row references.
using System;
using System.Buffers.Binary;
using System.IO;
using System.Text;

namespace Polygen.Common
{
    /// <summary>
    /// Owns the byte buffer behind binary row references.
    /// Row reference classes keep this object reachable, so the underlying buffer stays alive.
    /// </summary>
    public sealed class BinaryDocumentOwner
    {
        private readonly byte[] _buffer;

        public BinaryDocumentOwner(byte[] buffer)
        {
            _buffer = buffer ?? throw new ArgumentNullException(nameof(buffer));
        }

        public ReadOnlySpan<byte> Span => _buffer;

        public int Length => _buffer.Length;

        public static BinaryDocumentOwner Open(string filePath)
        {
            return new BinaryDocumentOwner(File.ReadAllBytes(filePath));
        }

        public BinaryReader OpenReader()
        {
            return new BinaryReader(new MemoryStream(_buffer, writable: false), Encoding.UTF8);
        }

        public BinaryReader OpenReaderAt(int offset)
        {
            BinaryRefFormat.CheckRange(_buffer, offset, 0);
            return new BinaryReader(
                new MemoryStream(_buffer, offset, _buffer.Length - offset, writable: false),
                Encoding.UTF8);
        }
    }

    public static class BinaryRefFormat
    {
        private static readonly byte[] Magic = { (byte)'P', (byte)'G', (byte)'B', (byte)'R', (byte)'E', (byte)'F', (byte)'1', 0 };
        private const int Version = 2;

        public static void WriteHeader(BinaryWriter writer)
        {
            writer.Write(Magic);
            writer.Write(Version);
        }

        public static void ReadHeader(BinaryReader reader)
        {
            var magic = reader.ReadBytes(Magic.Length);
            if (magic.Length != Magic.Length)
                throw new InvalidDataException("Invalid PolyGen binary ref header.");

            for (var i = 0; i < Magic.Length; i++)
            {
                if (magic[i] != Magic[i])
                    throw new InvalidDataException("Invalid PolyGen binary ref header.");
            }

            var version = reader.ReadInt32();
            if (version != Version)
                throw new InvalidDataException($"Unsupported PolyGen binary ref version: {version}.");
        }

        public static void WriteString(BinaryWriter writer, string value)
        {
            var bytes = Encoding.UTF8.GetBytes(value);
            writer.Write(bytes.Length);
            writer.Write(bytes);
        }

        public static string ReadString(BinaryReader reader)
        {
            var length = reader.ReadInt32();
            if (length < 0)
                throw new InvalidDataException("Negative string length.");
            var bytes = reader.ReadBytes(length);
            if (bytes.Length != length)
                throw new EndOfStreamException();
            return Encoding.UTF8.GetString(bytes);
        }

        public static int RequireFieldOffset(ReadOnlySpan<byte> buffer, int rowOffset, int fieldIndex)
        {
            var offset = GetFieldOffset(buffer, rowOffset, fieldIndex);
            if (offset < 0)
                throw new InvalidDataException($"Missing required binary field at index {fieldIndex}.");
            return offset;
        }

        public static int GetFieldOffset(ReadOnlySpan<byte> buffer, int rowOffset, int fieldIndex)
        {
            CheckRange(buffer, rowOffset, sizeof(int));
            var fieldCount = BinaryPrimitives.ReadInt32LittleEndian(buffer.Slice(rowOffset, sizeof(int)));
            if (fieldCount < 0)
                throw new InvalidDataException("Negative binary field count.");
            if (fieldIndex < 0 || fieldIndex >= fieldCount)
                return -1;

            var tableOffset = rowOffset + sizeof(int);
            CheckRange(buffer, tableOffset, checked(fieldCount * sizeof(int)));
            var relative = BinaryPrimitives.ReadInt32LittleEndian(buffer.Slice(tableOffset + fieldIndex * sizeof(int), sizeof(int)));
            if (relative < 0)
                return -1;

            var absolute = checked(rowOffset + relative);
            CheckRange(buffer, absolute, 0);
            return absolute;
        }

        public static ReadOnlySpan<byte> ReadLengthPrefixedBytes(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(int));
            var length = BinaryPrimitives.ReadInt32LittleEndian(buffer.Slice(offset, sizeof(int)));
            if (length < 0)
                throw new InvalidDataException("Negative binary payload length.");
            var payloadOffset = checked(offset + sizeof(int));
            CheckRange(buffer, payloadOffset, length);
            return buffer.Slice(payloadOffset, length);
        }

        public static string ReadUtf8String(ReadOnlySpan<byte> buffer, int offset)
        {
            return Encoding.UTF8.GetString(ReadLengthPrefixedBytes(buffer, offset));
        }

        public static bool ReadBool(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, 1);
            return buffer[offset] != 0;
        }

        public static byte ReadU8(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, 1);
            return buffer[offset];
        }

        public static sbyte ReadI8(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, 1);
            return unchecked((sbyte)buffer[offset]);
        }

        public static ushort ReadU16(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(ushort));
            return BinaryPrimitives.ReadUInt16LittleEndian(buffer.Slice(offset, sizeof(ushort)));
        }

        public static short ReadI16(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(short));
            return BinaryPrimitives.ReadInt16LittleEndian(buffer.Slice(offset, sizeof(short)));
        }

        public static uint ReadU32(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(uint));
            return BinaryPrimitives.ReadUInt32LittleEndian(buffer.Slice(offset, sizeof(uint)));
        }

        public static int ReadI32(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(int));
            return BinaryPrimitives.ReadInt32LittleEndian(buffer.Slice(offset, sizeof(int)));
        }

        public static ulong ReadU64(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(ulong));
            return BinaryPrimitives.ReadUInt64LittleEndian(buffer.Slice(offset, sizeof(ulong)));
        }

        public static long ReadI64(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(long));
            return BinaryPrimitives.ReadInt64LittleEndian(buffer.Slice(offset, sizeof(long)));
        }

        public static float ReadF32(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(float));
            return BinaryPrimitives.ReadSingleLittleEndian(buffer.Slice(offset, sizeof(float)));
        }

        public static double ReadF64(ReadOnlySpan<byte> buffer, int offset)
        {
            CheckRange(buffer, offset, sizeof(double));
            return BinaryPrimitives.ReadDoubleLittleEndian(buffer.Slice(offset, sizeof(double)));
        }

        public static TEnum ReadEnumInt32<TEnum>(ReadOnlySpan<byte> buffer, int offset) where TEnum : struct, Enum
        {
            var raw = ReadI32(buffer, offset);
            if (!Enum.IsDefined(typeof(TEnum), raw))
                throw new InvalidDataException($"Invalid enum value for {typeof(TEnum).FullName}: {raw}");
            return (TEnum)Enum.ToObject(typeof(TEnum), raw);
        }

        public static DateTime ReadTimestamp(ReadOnlySpan<byte> buffer, int offset)
        {
            return new DateTime(ReadI64(buffer, offset), DateTimeKind.Utc);
        }

        public static void CheckRange(ReadOnlySpan<byte> buffer, int offset, int length)
        {
            if (offset < 0 || length < 0 || offset > buffer.Length || length > buffer.Length - offset)
                throw new InvalidDataException("Binary ref offset is outside the document.");
        }
    }

    public sealed class BinaryRefRowBuilder
    {
        private readonly byte[]?[] _fields;

        public BinaryRefRowBuilder(int fieldCount)
        {
            if (fieldCount < 0)
                throw new ArgumentOutOfRangeException(nameof(fieldCount));
            _fields = new byte[fieldCount][];
        }

        public void SetField(int index, Action<BinaryWriter> write)
        {
            using var stream = new MemoryStream();
            using (var writer = new BinaryWriter(stream, Encoding.UTF8, leaveOpen: true))
            {
                write(writer);
            }
            _fields[index] = stream.ToArray();
        }

        public byte[] ToArray()
        {
            using var stream = new MemoryStream();
            using var writer = new BinaryWriter(stream, Encoding.UTF8, leaveOpen: true);

            writer.Write(_fields.Length);
            var headerSize = sizeof(int) + _fields.Length * sizeof(int);
            var cursor = headerSize;

            foreach (var field in _fields)
            {
                if (field == null)
                {
                    writer.Write(-1);
                }
                else
                {
                    writer.Write(cursor);
                    cursor = checked(cursor + field.Length);
                }
            }

            foreach (var field in _fields)
            {
                if (field != null)
                    writer.Write(field);
            }

            return stream.ToArray();
        }
    }
}
